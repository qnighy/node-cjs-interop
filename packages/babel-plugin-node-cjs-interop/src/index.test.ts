import { describe, expect, it } from "@jest/globals";
import { transform, TransformOptions } from "@babel/core";
import plugin from ".";

const defaultOptions: TransformOptions = {
  presets: [],
  plugins: [[plugin, { modulePrefixes: ["mod"] }]],
  configFile: false,
  babelrc: false,
};

describe("babel-plugin-node-cjs-interop", () => {
  it("wraps default imports", () => {
    const result = transform(`import f from "mod";\nconsole.log(f);\n`, defaultOptions);
    expect(result?.code).toMatchSnapshot();
  });

  it("wraps named default imports", () => {
    const result = transform(`import { default as f } from "mod";\nconsole.log(f);\n`, defaultOptions);
    expect(result?.code).toMatchSnapshot();
  });

  it("wraps namespace imports", () => {
    const result = transform(`import * as M from "mod";\nconsole.log(M);\n`, defaultOptions);
    expect(result?.code).toMatchSnapshot();
  });

  it("doesn't wrap named imports", () => {
    const result = transform(`import { f } from "mod";\nconsole.log(f);\n`, defaultOptions);
    expect(result?.code).toMatchSnapshot();
  });

  describe("with typescript", () => {
    const options: TransformOptions = {
      ...defaultOptions,
      presets: ["@babel/preset-typescript"],
      filename: "test.ts",
    };

    it("wraps default imports", () => {
      const result = transform(`import f from "mod";\nconsole.log(f);\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("wraps named default imports", () => {
      const result = transform(`import { default as f } from "mod";\nconsole.log(f);\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("wraps namespace imports", () => {
      const result = transform(`import * as M from "mod";\nconsole.log(M);\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("doesn't wrap named imports", () => {
      const result = transform(`import { f } from "mod";\nconsole.log(f);\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("doesn't wrap type imports (default imports)", () => {
      const result = transform(`import type f from "mod";\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("doesn't wrap type imports (named default imports)", () => {
      const result = transform(`import type { default as f } from "mod";\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("doesn't wrap type imports (named default imports, individual specification)", () => {
      const result = transform(`import { type default as f } from "mod";\n`, options);
      expect(result?.code).toMatchSnapshot();
    });

    it("doesn't wrap type imports (namespace imports)", () => {
      const result = transform(`import type * as M from "mod";\n`, options);
      expect(result?.code).toMatchSnapshot();
    });
  });

  describe("source filtering", () => {
    it("filters sources", () => {
      const result = transform(`import foo from "foo";\nimport bar from "bar";\nconsole.log(foo, bar)\n`, {
        ...defaultOptions,
        plugins: [[plugin, { modulePrefixes: ["foo"] }]],
      });
      expect(result?.code).toMatchSnapshot();
    });

    it("filters multiple sources", () => {
      const result = transform(`import foo from "foo";\nimport bar from "bar";\nimport baz from "baz";\nconsole.log(foo, bar, baz)\n`, {
        ...defaultOptions,
        plugins: [[plugin, { modulePrefixes: ["foo", "bar"] }]],
      });
      expect(result?.code).toMatchSnapshot();
    });

    it("matches subdirectories", () => {
      const result = transform(`import foo from "foo";\nimport bar from "foo/bar";\nimport baz from "baz";\nconsole.log(foo, bar, baz)\n`, {
        ...defaultOptions,
        plugins: [[plugin, { modulePrefixes: ["foo"] }]],
      });
      expect(result?.code).toMatchSnapshot();
    });

    it("doesn't match non-subdirectory prefixes", () => {
      const result = transform(`import foo from "foo";\nimport bar from "foobar";\nimport baz from "baz";\nconsole.log(foo, bar, baz)\n`, {
        ...defaultOptions,
        plugins: [[plugin, { modulePrefixes: ["foo"] }]],
      });
      expect(result?.code).toMatchSnapshot();
    });

    it("allows specifying scoped packages", () => {
      const result = transform(`import foo from "@foo/foo";\nimport bar from "@foo/bar";\nimport baz from "baz";\nconsole.log(foo, bar, baz)\n`, {
        ...defaultOptions,
        plugins: [[plugin, { modulePrefixes: ["@foo/bar"] }]],
      });
      expect(result?.code).toMatchSnapshot();
    });
  });
})
