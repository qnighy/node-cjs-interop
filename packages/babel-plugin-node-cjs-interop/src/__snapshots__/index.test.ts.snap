// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-node-cjs-interop doesn't wrap named imports 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(_ns.f);"
`;

exports[`babel-plugin-node-cjs-interop source filtering allows specifying scoped packages 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

import foo from \\"@foo/foo\\";

/*#__CJS__*/
import * as _nsOrig from \\"@foo/bar\\";
import baz from \\"baz\\";
console.log(foo, _ns.default, baz);"
`;

exports[`babel-plugin-node-cjs-interop source filtering doesn't match non-subdirectory prefixes 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"foo\\";
import bar from \\"foobar\\";
import baz from \\"baz\\";
console.log(_ns.default, bar, baz);"
`;

exports[`babel-plugin-node-cjs-interop source filtering filters multiple sources 1`] = `
"const _ns2 = _interopImportCJSNamespace(_nsOrig2);

const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"foo\\";

/*#__CJS__*/
import * as _nsOrig2 from \\"bar\\";
import baz from \\"baz\\";
console.log(_ns.default, _ns2.default, baz);"
`;

exports[`babel-plugin-node-cjs-interop source filtering filters sources 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"foo\\";
import bar from \\"bar\\";
console.log(_ns.default, bar);"
`;

exports[`babel-plugin-node-cjs-interop source filtering matches subdirectories 1`] = `
"const _ns2 = _interopImportCJSNamespace(_nsOrig2);

const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"foo\\";

/*#__CJS__*/
import * as _nsOrig2 from \\"foo/bar\\";
import baz from \\"baz\\";
console.log(_ns.default, _ns2.default, baz);"
`;

exports[`babel-plugin-node-cjs-interop with typescript doesn't wrap named imports 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(_ns.f);"
`;

exports[`babel-plugin-node-cjs-interop with typescript doesn't wrap type imports (default imports) 1`] = `"export {};"`;

exports[`babel-plugin-node-cjs-interop with typescript doesn't wrap type imports (named default imports) 1`] = `"export {};"`;

exports[`babel-plugin-node-cjs-interop with typescript doesn't wrap type imports (named default imports, individual specification) 1`] = `"export {};"`;

exports[`babel-plugin-node-cjs-interop with typescript doesn't wrap type imports (namespace imports) 1`] = `"export {};"`;

exports[`babel-plugin-node-cjs-interop with typescript wraps default imports 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(_ns.default);"
`;

exports[`babel-plugin-node-cjs-interop with typescript wraps named default imports 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(_ns.default);"
`;

exports[`babel-plugin-node-cjs-interop with typescript wraps namespace imports 1`] = `
"const M = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(M);"
`;

exports[`babel-plugin-node-cjs-interop wraps default imports 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(_ns.default);"
`;

exports[`babel-plugin-node-cjs-interop wraps named default imports 1`] = `
"const _ns = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(_ns.default);"
`;

exports[`babel-plugin-node-cjs-interop wraps namespace imports 1`] = `
"const M = _interopImportCJSNamespace(_nsOrig);

function _interopImportCJSNamespace(ns) {
  return ns.default && ns.default.__esModule ? ns.default : ns;
}

/*#__CJS__*/
import * as _nsOrig from \\"mod\\";
console.log(M);"
`;
